src/main/java/org/terasology/logic/health/HealthAuthoritySystem.java

```diff
@@ -140,16 +165,199 @@ private int calculateTotal(int base, TFloatList multipliers, TIntList modifiers)
 
     }
 
-    /** Handles DoHeal event to increase health of entity */
+    /**
+     * Handles DoHeal event to increase health of entity.
+     *
+     * @param event DoHealEvent which triggered the heal effect.
+     * @param entity The entity which will be healed by the event.
+     */
     @ReceiveEvent(components = {HealthComponent.class})
     public void onHeal(DoHealEvent event, EntityRef entity) {
         checkHeal(entity, event.getAmount(), event.getInstigator());
     }
 
-    /** Resets the health of player to maxHealth on re-spawn. */
+    /**
+     * Resets the health of player to maxHealth on re-spawn.
+     *
+     * @param event OnPlayerRespawnedEvent sent during the re-spwan of player.
+     * @param entity The Player entity.
+     * @param healthComponent The health component attached to player entity.
+     */
     @ReceiveEvent
     public void onRespawn(OnPlayerRespawnedEvent event, EntityRef entity, HealthComponent healthComponent) {
         healthComponent.currentHealth = healthComponent.maxHealth;
         entity.saveComponent(healthComponent);
     }
+
+
+    /**
+     * Override the default behavior for an attack, causing it damage as opposed to just destroying it or doing nothing.
+     *
+     * @param event Attack event sent on targetEntity.
+     * @param targetEntity The entity which is attacked.
+     */
+    @ReceiveEvent(components = HealthComponent.class, netFilter = RegisterMode.AUTHORITY)
+    public void onAttackEntity(AttackEvent event, EntityRef targetEntity) {
+        damageEntity(event, targetEntity);
+    }
+
+    static void damageEntity(AttackEvent event, EntityRef targetEntity) {
+        int damage = 1;
+        Prefab damageType = EngineDamageTypes.PHYSICAL.get();
+        // Calculate damage from item
+        ItemComponent item = event.getDirectCause().getComponent(ItemComponent.class);
+        if (item != null) {
+            damage = item.baseDamage;
+            if (item.damageType != null) {
+                damageType = item.damageType;
+            }
+        }
+
+        targetEntity.send(new DoDamageEvent(damage, damageType, event.getInstigator(), event.getDirectCause()));
+        // consume the event so that the health system can take priority over default engine behavior
+        event.consume();
+    }
+
+    private void doDamage(EntityRef entity, int damageAmount, Prefab damageType, EntityRef instigator, EntityRef directCause) {
+        HealthComponent health = entity.getComponent(HealthComponent.class);
+        CharacterMovementComponent characterMovementComponent = entity.getComponent(CharacterMovementComponent.class);
+        boolean ghost = false;
+        if (characterMovementComponent != null) {
+            ghost = (characterMovementComponent.mode == MovementMode.GHOSTING);
+        }
+        if ((health != null) && !ghost) {
+            int damagedAmount = health.currentHealth - Math.max(health.currentHealth - damageAmount, 0);
```

Comment:
This is overly complicated and the name is ambiguous.
Can't you replace this line with
```java
int damagedAmount = Math.min(health.currentHealth, damageAmount)
```
Further more based on that it makes sense to use `maxDamage` or `cappedDamage` as the name, eliminating the confusion between `damageAmount` and `damagedAmount`

Reply:
This also applies to a similar section of the `doHeal` method